pkg/cloudprovider/aws/fleet/capacity.go

```diff
@@ -0,0 +1,183 @@
+	// 1. Select a zone
+	zone, err := c.selectZone(ctx, constraints)
+	if err != nil {
+		return nil, fmt.Errorf("getting zone, %w", err)
+	}
+
+	// 2. Select a zubnet, limited to selected zone
+	subnet, err := c.selectSubnet(ctx, zone, constraints)
+	if err != nil {
+		return nil, fmt.Errorf("getting subnet, %w", err)
+	}
+
+	// 3. Detect launch template
+	launchTemplate, err := c.launchTemplateProvider.Get(ctx, c.spec.Cluster)
+	if err != nil {
+		return nil, fmt.Errorf("getting launch template, %w", err)
+	}
+
+	// 3. Create Fleet
+	createFleetOutput, err := c.ec2.CreateFleetWithContext(ctx, &ec2.CreateFleetInput{
+		Type: aws.String(ec2.FleetTypeInstant),
+		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
+			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeOnDemand), // TODO support SPOT
+			TotalTargetCapacity:       aws.Int64(1), // TODO construct this more intelligently
+		},
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{{
+			LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+				LaunchTemplateName: launchTemplate.LaunchTemplateName,
+				Version:            aws.String("$Default"),
+			},
+			Overrides: []*ec2.FleetLaunchTemplateOverridesRequest{{
+				AvailabilityZone: aws.String(zone),
+				InstanceType:     aws.String("m5.large"), // TODO construct this more intelligently
+				SubnetId:         subnet.SubnetId,
+			}},
+		}},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("creating fleet, %w", err)
+	}
+
+	// 4. Transform to Nodes
+	var instanceIds []*string
+	for _, fleetInstance := range createFleetOutput.Instances {
+		instanceIds = append(instanceIds, fleetInstance.InstanceIds...)
+	}
+	nodes, err := c.nodeFactory.For(ctx, instanceIds)
+	if err != nil {
+		return nil, fmt.Errorf("determining nodes, %w", err)
+	}
+	zap.S().Infof("Successfully requested %d nodes", len(nodes))
+	return nodes, nil
+}
+
+// GetTopologyDomains returns a set of supported domains.
+// e.g. us-west-2 -> [ us-west-2a, us-west-2b ]
+func (c *Capacity) GetTopologyDomains(ctx context.Context, key cloudprovider.TopologyKey) ([]string, error) {
+	switch key {
+	case cloudprovider.TopologyKeyZone:
+		zones, err := c.getZones(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return zones, nil
+	case cloudprovider.TopologyKeySubnet:
+		subnets, err := c.getSubnetIds(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return subnets, nil
+	default:
+		return nil, fmt.Errorf("unrecognized topology key %s", key)
+	}
+}
+
+// seletZone chooses a zone for the given constraints.
+func (c *Capacity) selectZone(ctx context.Context, constraints *cloudprovider.CapacityConstraints) (string, error) {
+	// 1. Return zone if specified
+	if zone, ok := constraints.Topology[cloudprovider.TopologyKeyZone]; ok {
+		return zone, nil
+	}
+	// 2. Randomly choose from available zones
+	zones, err := c.getZones(ctx)
+	if err != nil {
+		return "", err
+	}
+	return zones[rand.Intn(len(zones))], nil
```

Comment:
Is `Seed` called somewhere?

Reply:
No. IIUC, it's only necessary if you want determinism: https://golang.org/pkg/math/rand/#Seed

Reply:
Oh I'm super wrong. Will fix.